### PR TITLE
Add check to see if testdata is empty when determining when to add data

### LINKED ISSUE #2233: Gazelle adds data source even for empty testdata directory
Recently ran into an issue where we removed a testdata directory in favor of a different testing approach. In the aftermath of this change we had a couple PRs that attempted to add `data = glob(["testdata/**"]),` to our BUILD.bazel. This was due to the presence of an untracked empty testdata directory after rebase.

I can't seem to think of a scenario in which we would want an empty testdata directory to be added as a data src. I believe the change could be narrowed down to adding a check for an empty subdir in the following code.

https://github.com/bazel-contrib/bazel-gazelle/blob/b13c1573c378bb3dd9b9de1397db89d0392f8e72/language/go/generate.go#L113-L115

### PR DESCRIPTION
**Pull request by sohumk-stripe**
**What type of PR is this?**

Bug fix

**What package or component does this PR mostly affect?**

language/go

**What does this PR do? Why is it needed?**

Adds logic to check that a testdata subdirectory is not empty before adding it as a data source.

**Which issues(s) does this PR fix?**

Fixes #2233

### CODE REVIEW

File: language/go/generate.go [hunk 1]
```diff
@@ -107,12 +108,21 @@ func (gl *goLang) GenerateRules(args language.GenerateArgs) language.GenerateRes
...
 	var hasTestdata bool
 	for _, sub := range args.Subdirs {
 		if sub == "testdata" {
 			_, ok := gl.goPkgRels[path.Join(args.Rel, "testdata")]
 			hasTestdata = !ok
+
+			// Check if testdata directory is empty
+			if hasTestdata {
+				testdataPath := filepath.Join(args.Dir, "testdata")
+				entries, err := os.ReadDir(testdataPath)
```
Comment (jayconrod) [inline language/go/generate.go]: Use [`walk.GetDirInfo`](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle/walk#GetDirInfo) to read directory contents rather than direct I/O. The `walk` package already caches this in memory.

File: language/go/generate.go [hunk 2]
```diff
@@ -107,12 +108,21 @@ func (gl *goLang) GenerateRules(args language.GenerateArgs) language.GenerateRes
...
-	// not contain a buildable package.
+	// not contain a buildable package and is not empty.
 	var hasTestdata bool
 	for _, sub := range args.Subdirs {
 		if sub == "testdata" {
 			_, ok := gl.goPkgRels[path.Join(args.Rel, "testdata")]
 			hasTestdata = !ok
+
+			// Check if testdata directory is empty
+			if hasTestdata {
```
Comment (jayconrod) [inline language/go/generate.go]: Nit: maybe just check `!ok` here instead of assigning `hasTestdata` above? It's a little confusing that `hasTestdata` is assigned more than once with different meaning.

File: language/go/generate.go (hunk 1 above)
Comment (sohumk-stripe) [reply language/go/generate.go]: We're not in a walk here, so I don't believe this would work?

File: language/go/generate.go (hunk 2 above)
Comment (sohumk-stripe) [reply language/go/generate.go]: Done, and also added a test.

File: language/go/generate.go (hunk 1 above)
Comment (jayconrod) [reply language/go/generate.go]: There's a hack in place to make it work through global state. (We'll improve that for v2, but don't worry about it for now.)

Comment (sohumk-stripe) [reply language/go/generate.go]: Can you share an example of that workaround?

Comment (jayconrod) [reply language/go/generate.go]: It's in https://github.com/bazel-contrib/bazel-gazelle/blob/master/walk/cache.go#L104, `globalWalker` is set by `Walk2`.

I realize that may not be active in the unit test, but `GenerateRules` is normally only called by `Walk2`.

Did you try adding a test in `//tests` instead? I imagine that might run into a problem with `git` not creating empty directories. `cmd/gazelle/integration_test.go` might be another place we could test this.

Comment (sohumk-stripe) [reply language/go/generate.go]: Okay, I see what you mean. My one concern here is whether we're certain `GenerateRules` is only used within a walk, as it would break if not.

Will look at adding a test in `//tests` and/or `cmd/gazelle/integration_test.go`

Comment (sohumk-stripe) [reply language/go/generate.go]: I updated the unit test to use Walk and changed the `os.ReadDir` to `walk.GetDirInfo`.

Still working on adding a test in `//tests` or `cmd/gazelle/integration_test.go`

Comment (sohumk-stripe) [reply language/go/generate.go]: Added an integration test, couldn't add anything to `//tests` that could be checked in because of `git` not tracking directories.

**Review by jayconrod — APPROVED** [review]
Thanks, this looks good.

I appreciate your patience with `Walk` and with the tests. Reducing I/O should speed things up. I'm working on Gazelle v2, and fixing `walk.GetDirInfo` will be part of that (among many other things).
